### PR TITLE
Unify handling of transforms

### LINKED ISSUE #1813: New `transforms` option doesn’t instantiate transform
In #1786, y’all kindly added a new `transform` options to the cli’s `main()` because of our discussion in torch2424/as-bind#75. However, I didn’t realize that the new code path doesn’t actually instantiate the transform or fixes up the prototype.

Transforms passed via `--transform` get this code path:
https://github.com/AssemblyScript/assemblyscript/blob/7d0690d91877a43a000bf17035c08acecaa93a25/cli/asc.js#L519-L529

while the new `main()` parameter gets:

https://github.com/AssemblyScript/assemblyscript/blob/7d0690d91877a43a000bf17035c08acecaa93a25/cli/asc.js#L505

I think it’d be great if both ways to pass in transforms get the same fix-ups applied.

### PR DESCRIPTION
**Pull request by surma**
Fixes #1813 

- [x] I've read the contributing guidelines

### CODE REVIEW

File: cli/asc.js [hunk 1]
```diff
@@ -514,28 +517,35 @@ exports.main = function main(argv, options, callback) {
...
+          program,
+          baseDir,
+          stdout,
+          stderr,
+          log: console.error,
+          readFile,
+          writeFile,
+          listFiles
+        });
+    });
```
Comment (MaxGraey) [inline cli/asc.js:544]:
```suggestion
      return Object.assign(classOrModule.prototype, {
        program,
        baseDir,
        stdout,
        stderr,
        log: console.error,
        readFile,
        writeFile,
        listFiles
      });
    });
```

File: cli/asc.js [hunk 2]
```diff
@@ -514,28 +517,35 @@ exports.main = function main(argv, options, callback) {
...
-            writeFile,
-            listFiles
-          });
-          transforms.push(new classOrModule());
-        } else {
-          transforms.push(classOrModule); // legacy module
-        }
-      } catch (e) {
+        transforms.push(dynrequire(dynrequire.resolve(filename, { paths: [baseDir, process.cwd()] })));
+      } catch(e) {
```
Comment (MaxGraey) [inline cli/asc.js:521]:
```suggestion
      } catch (e) {
```

**Comment by surma** [comment]
Cheers @MaxGraey

File: cli/asc.js [hunk 3]
```diff
@@ -514,28 +517,35 @@ exports.main = function main(argv, options, callback) {
...
+        stdout,
+        stderr,
+        log: console.error,
+        readFile,
+        writeFile,
+        listFiles
+      });
+    });
+  } catch (e) {
+    return callback(e)
```
Comment (MaxGraey) [inline cli/asc.js:546]:
```suggestion
    return callback(e);
```
last lint error it seems

File: cli/asc.js [hunk 4]
```diff
@@ -499,11 +499,14 @@ exports.main = function main(argv, options, callback) {
   // Initialize the program
   program = __pin(assemblyscript.newProgram(compilerOptions));
 
-  // Set up transforms
-  const transforms = [];
+  // Collect transforms *constructors* from the `--transform` CLI flag as well
+  // as the `transform` option into the `transforms` array.
+  let transforms = [];
+  // `transform` option from `main()`
   if (Array.isArray(options.transforms)) {
```
Comment (MaxGraey) [inline cli/asc.js:506]: I'm also wondering is normal that `options.transforms` and `opts.transform` [here](https://github.com/AssemblyScript/assemblyscript/pull/1814/files#diff-609c82113feed36cb75ab2bd3d387762bf2829fc5ae10477722fee8a97885489R510) have different key names? I'm not familiar with all this option keys

Comment (surma) [reply cli/asc.js:506]: Yeah, it is confusing (but as you can tell, that was already the case in the old code). I wanted to keep this PR minimal and not start renaming a lot of things.

`options` is the options object passed to `main()`, while `opts` is the return value of the command line flag parser. That’s why I added comments that one of the branches is about handling the transforms from the options object, while the other one handles transforms from the CLI flag.

File: cli/asc.js (hunk 3 above)
Comment (surma) [reply cli/asc.js:546]: I need to check why I didn’t catch these locally...

**Review by MaxGraey — APPROVED** [review]
LGTM. Could you confirm does it work?

**Comment by surma** [comment]
@MaxGraey I wanted to, but I am honestly not sure how to do that as [the SDK hard-codes CDN URLs](https://github.com/AssemblyScript/assemblyscript/blob/7d0690d91877a43a000bf17035c08acecaa93a25/lib/sdk/index.js#L7-L12). It’s something that I think should be fixed, but probably in a separate PR.

**Comment by MaxGraey** [comment]
Thanks!

**Pull request merged**